python/cudf/cudf/core/series.py

```diff
@@ -536,11 +536,24 @@ def __init__(
                 data = data.astype(dtype)
 
         if isinstance(data, dict):
-            index = data.keys()
-            data = column.as_column(
-                list(data.values()), nan_as_null=nan_as_null, dtype=dtype
-            )
-
+            current_index = data.keys()
+            if index is not None:
+                series = Series(
+                    list(data.values()),
+                    nan_as_null=nan_as_null,
+                    dtype=dtype,
+                    index=current_index,
+                )
+                new_index = as_index(index)
+                if not series.index.equals(new_index):
+                    series = series.reindex(new_index)
+                data = series._column
+                index = series._index
```

Comment:
Could you explain why this block is needed? Does the reindexing allow you to provide mismatched index and data lengths as long as they are commensurate?

Reply:
We need this block of code to support over-riding the constructed dataframe's `index`:
previously, it did not work:
```python
In [4]: s = cudf.Series({'a':1, 'b':10, 'c':20}, index=['c', 'b'])

In [5]: s
Out[5]: 
a     1
b    10
c    20
dtype: int64
```

This block of code fixes the issue:
```python
In [7]: s = cudf.Series({'a':1, 'b':10, 'c':20}, index=['c', 'b'])

In [8]: s
Out[8]: 
c    20
b    10
dtype: int64
```

Reply:
Yep OK that makes sense.